tar/lib/src/utils.dart

```diff
@@ -0,0 +1,680 @@
+// Copyright 2020
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import 'dart:convert';
+import 'dart:math';
+import 'dart:io';
+
+import 'package:path/path.dart' as p;
+import 'package:tar/src/header.dart';
+
+import 'constants.dart';
+import 'exceptions.dart';
+import 'sparse_entry.dart';
+
+/// Parses the sublist of [bytes] from [start] to [end] as a String. First it
+/// checks if it is utf8-encoded, and if that fails, we revert to reading it
+/// byte by byte.
+String parseString(List<int> bytes, [int start, int end]) {
+  ArgumentError.checkNotNull(bytes, 'bytes');
+
+  start ??= 0;
+  end ??= bytes.length;
+
+  RangeError.checkValidIndex(start, bytes);
+  RangeError.checkValidIndex(end - 1, bytes);
+
+  final rawCharCodes = bytes.sublist(start, end);
+
+  final stringEnd = rawCharCodes.indexOf(NUL);
+  final charCodes = rawCharCodes.sublist(0, stringEnd == -1 ? null : stringEnd);
+
+  try {
+    return utf8.decode(charCodes);
+  } catch (e) {
```

Comment:
I believe we should only catch FormatErrors here.
```suggestion
  } on FormatError catch (e) {
```